File: backend/app/schemas/meta.py
"""Meta-progression schemas (essence, talents, achievements, lore)."""

from pydantic import BaseModel


class MetaResponse(BaseModel):
    essence_balance: int
    essence_lifetime: int
    unlocked_talents: list[str]
    achievements: list[str]
    total_runs_completed: int
    total_runs_won: int
    highest_floor_reached: int
    total_floors_cleared: int
    total_deaths_all_runs: int
    secret_floors_discovered: list[str]
    lore_fragments_found: list[str]
    antagonist_encounters: int


class TalentResponse(BaseModel):
    id: str
    name: str
    branch: str
    tier: int
    cost: int
    effect_key: str
    description: str
    is_unlocked: bool
    can_afford: bool
    prerequisite_met: bool


class TalentTreeResponse(BaseModel):
    talents: list[TalentResponse]
    essence_balance: int


class AchievementResponse(BaseModel):
    id: str
    name: str
    description: str
    category: str
    essence_reward: int
    is_earned: bool


class LoreFragmentResponse(BaseModel):
    id: str
    title: str
    text: str
    category: str
    source: str
    is_discovered: bool


class LoreBeatResponse(BaseModel):
    id: str
    floor: int
    act: int
    trigger: str
    arbiter_text: str
    aethon_text: str | None
    dm_stage_direction: str
    lore_fragment_id: str | None


class RunEndMetaResponse(BaseModel):
    essence_earned: int
    new_achievements: list[AchievementResponse]
    lore_fragments_discovered: list[LoreFragmentResponse]
    total_essence: int
    talents_affordable: int

File: frontend/src/api/meta.ts
import { api } from "./client";

// ── Response Types ──────────────────────────────────────────────────────────

export interface MetaResponse {
  essence_balance: number;
  essence_lifetime: number;
  unlocked_talents: string[];
  achievements: string[];
  total_runs_completed: number;
  total_runs_won: number;
  highest_floor_reached: number;
  total_floors_cleared: number;
  total_deaths_all_runs: number;
  secret_floors_discovered: string[];
  lore_fragments_found: string[];
  antagonist_encounters: number;
}

export interface TalentResponse {
  id: string;
  name: string;
  branch: string;
  tier: number;
  cost: number;
  effect_key: string;
  description: string;
  is_unlocked: boolean;
  can_afford: boolean;
  prerequisite_met: boolean;
}

export interface TalentTreeResponse {
  talents: TalentResponse[];
  essence_balance: number;
}

export interface AchievementResponse {
  id: string;
  name: string;
  description: string;
  category: string;
  essence_reward: number;
  is_earned: boolean;
}

export interface LoreFragmentResponse {
  id: string;
  title: string;
  text: string;
  category: string;
  source: string;
  is_discovered: boolean;
}

export interface LoreBeatResponse {
  id: string;
  floor: number;
  act: number;
  trigger: string;
  arbiter_text: string;
  aethon_text: string;
  dm_stage_direction: string;
  lore_fragment_id: string | null;
}

export interface RunEndMetaResponse {
  essence_earned: number;
  new_achievements: AchievementResponse[];
  lore_fragments_discovered: LoreFragmentResponse[];
  total_essence: number;
  talents_affordable: number;
}

// ── API Client ──────────────────────────────────────────────────────────────

export const metaApi = {
  /** Get campaign meta-progression overview (auto-creates if missing) */
  get: (campaignId: string) =>
    api.get<MetaResponse>(`/campaigns/${campaignId}/meta`),

  /** Get the full talent tree with unlock/affordability status */
  getTalents: (campaignId: string) =>
    api.get<TalentTreeResponse>(`/campaigns/${campaignId}/meta/talents`),

  /** Spend essence to unlock a talent */
  unlockTalent: (campaignId: string, talentId: string) =>
    api.post<TalentTreeResponse>(
      `/campaigns/${campaignId}/meta/talents/${talentId}/unlock`,
    ),

  /** Get all achievements with earned/unearned status */
  getAchievements: (campaignId: string) =>
    api.get<AchievementResponse[]>(`/campaigns/${campaignId}/meta/achievements`),

  /** Get all lore fragments with discovered/undiscovered status */
  getLore: (campaignId: string) =>
    api.get<LoreFragmentResponse[]>(`/campaigns/${campaignId}/meta/lore`),

  /** Get lore beats for a specific floor and trigger */
  getLoreBeats: (campaignId: string, floorNumber: number, trigger = "floor_end") =>
    api.get<LoreBeatResponse[]>(
      `/campaigns/${campaignId}/meta/lore-beats/${floorNumber}?trigger=${trigger}`,
    ),

  /** Complete a run and compute meta-progression rewards */
  completeRunMeta: (campaignId: string, runId: string) =>
    api.post<RunEndMetaResponse>(
      `/campaigns/${campaignId}/runs/${runId}/complete-meta`,
    ),
};
